### PR TITLE
ENH: Add ordered QZ decomposition

### PR DESCRIPTION
**Pull request by jseabold**
This currently segfaults. I don't know why. I spent some time checking the wrapper definitions and stepping through with gdb. I didn't notice anything that was obvious to me. If someone else wants to give it a go, this was the working script I was using. If someone can figure out the segfault, I'll take it the rest of the way, but I just don't have time to mess with this.

```
A = [[-21.10-22.50j, 53.5-50.5j, -34.5+127.5j,   7.5+ 0.5j],
[ -0.46- 7.78j, -3.5-37.5j, -15.5+ 58.5j, -10.5- 1.5j],
[  4.30- 5.50j, 39.7-17.1j, -68.5+ 12.5j,  -7.5- 3.5j],
[  5.50+ 4.40j, 14.4+43.3j, -32.5- 46.0j, -19.0-32.5j]]
B = [[1.0-5.0j,  1.6+1.2j, -3+0j,  0.0-1.0j],
[0.8-0.6j,  3.0-5.0j, -4+3j, -2.4-3.2j],
[1.0+0.0j,  2.4+1.8j, -4-5j,  0.0-3.0j],
[0.0+1.0j, -1.8+2.4j,  0-4j,  4.0-5.0j]]

A = [[3.9, 12.5, -34.5, -0.5],
[4.3, 21.5, -47.5,  7.5],
[4.3, 21.5, -43.5,  3.5],
[4.4, 26.0, -46.0,  6.0]]

B = [[  1,    2,   -3,    1],
[  1,    3,   -5,    4],
[  1,    3,   -4,    3],
[  1,    3,   -4,    4]]


from scipy.linalg._decomp_qz import ordqz, _qz, _select_function, get_lapack_funcs
output = 'real'
lwork = None
overwrite_a = False
overwrite_b = False
check_finite = True
sort = 'lhp'
result, typ = _qz(A, B, output=output, lwork=lwork, sort=None,
                 overwrite_a=overwrite_a, overwrite_b=overwrite_b,
                 check_finite=check_finite)
AA, BB, Q, Z = result[0], result[1], result[-4], result[-3]
alpha, beta = result[3], result[4]
sfunction = _select_function(sort, typ)
select = sfunction(alpha, beta)
tgsen, = get_lapack_funcs(('tgsen',), (AA,BB))

n = len(A)
m = 0.
lwork = 2*m*(n-m)
liwork = n+2
#result = tgsen(select, AA, BB, Q, Z, lwork=-1, liwork=AA.shape[1]*2)
result = tgsen(select, AA, BB, Q, Z)
```

### CODE REVIEW

**Comment by ricardomayerb** [comment]
Since the work to put together release 0.16.0 it's about to start, I humbly ask if there any chance to resolve this QZ issue for this upcoming release.

Also, it could be useful if someone could state what are the skills needed to resolve this, because if no one has the time to work on this I'd be interested in contributing to it, but I suspect there is a lot I would  have to learn before I can be of any use.

**Comment by jseabold** [comment]
Rebased. To go in, someone needs to debug the issue that apparently only shows up on Python 3.2 32-bit windows unless we've since updated our ATLAS/LAPACK and the problem has gone away.

**Comment by rgommers** [comment]
``pep8 scipy`` is still quite unhappy

**Comment by rgommers** [comment]
I'm right now messing with the Windows ATLAS binaries, because they're missing some routines that were wrapped for 0.16.0. I'll have a look at this again.

**Comment by jseabold** [comment]
Travis is happy now. Let me know anything else I can do.

File: scipy/linalg/flapack.pyf.src [hunk 1]
```diff
@@ -16,9 +16,71 @@ interface
...
+     integer optional,intent(in),depend(m,n),check(lwork == -1 || lwork >= 2*m*(n-m)) :: lwork=2*m*(n-m)
+     integer intent(out),dimension(MAX(1,liwork)):: iwork
+     integer optional,intent(in),depend(n),check(liwork == -1 || liwork>=n+2) :: liwork=n+2
+     integer intent(out) :: info
+   end subroutine <prefix2c>tgsen
+   
    subroutine <prefix2>gges(jobvsl,jobvsr,sort_t,<prefix2>select,n,a,lda,b,ldb,sdim,alphar,alphai,beta,vsl,ldvsl,vsr,ldvsr,work,lwork,bwork,info)
 
-     ! For a pair of N-by-N real nonsymmetric matrices (A,B) computes
+     ! For a pair of N-by-N real nonsymetric matrices (A,B) computes
```
Comment (rgommers) [inline scipy/linalg/flapack.pyf.src]: introducing a typo here, can you revert?

File: scipy/linalg/tests/test_decomp.py [hunk 2]
```diff
@@ -1946,22 +1946,181 @@ def test_qz_double_sort(self):
...
 
 
+def _make_pos(X):
+    # the decompositions can have different signs than verified results
+    return np.sign(X)*X
+
+
+class TestOrdQZ(TestCase):
+    @classmethod
+    def setupClass(cls):
```
Comment (rgommers) [inline scipy/linalg/tests/test_decomp.py]: Any reason not to change the above two lines to the standard ``def setUp(self):``?

Comment (rgommers) [inline scipy/linalg/tests/test_decomp.py]: and the ``cls.`` below to ``self.``

File: scipy/linalg/_decomp_qz.py [hunk 3]
```diff
@@ -10,45 +10,102 @@
...
+
+    return sfunction
+
+
+def _qz(A, B, output='real', lwork=None, sort=None, overwrite_a=False,
+        overwrite_b=False, check_finite=True):
+    if sort is not None:
+        # Disabled due to segfaults on win32, see ticket 1717.
+        raise ValueError("The 'sort' input of qz() has to be None and will be "
+                         "removed in 0.15. Use ordqz instead.")
```
Comment (rgommers) [inline scipy/linalg/_decomp_qz.py]: let's make this ``a future release`` instead of ``0.15``

File: scipy/linalg/_decomp_qz.py [hunk 4]
```diff
@@ -10,45 +10,102 @@
...
+        warnings.warn("The QZ iteration failed. (a,b) are not in Schur "
+                      "form, but ALPHAR(j), ALPHAI(j), and BETA(j) should be "
+                      "correct for J=%d,...,N" % info-1, UserWarning)
+    elif info == a_n + 1:
+        raise LinAlgError("Something other than QZ iteration failed")
+    elif info == a_n + 2:
+        raise LinAlgError("After reordering, roundoff changed values of some "
+                          "complex eigenvalues so that leading eigenvalues "
+                          "in the Generalized Schur form no longer satisfy "
+                          "sort=True. This could also be caused due to "
```
Comment (rgommers) [inline scipy/linalg/_decomp_qz.py]: remove ``caused`` here

File: scipy/linalg/_decomp_qz.py [hunk 5]
```diff
@@ -160,86 +214,139 @@ def qz(A, B, output='real', lwork=None, sort=None, overwrite_a=False,
            [-0.79813178,  0.58842606,  0.12938478],
            [-0.54861681, -0.6210585 , -0.55973739]])
 
+    See also
+    --------
+    scipy.linalg.ordqz
```
Comment (rgommers) [inline scipy/linalg/_decomp_qz.py]: simply ``ordqz`` is enough

File: scipy/linalg/_decomp_qz.py [hunk 6]
```diff
@@ -160,86 +214,139 @@ def qz(A, B, output='real', lwork=None, sort=None, overwrite_a=False,
...
+    A : (N, N) array_like
+        2d array to decompose
+    B : (N, N) array_like
+        2d array to decompose
+    sort : {callable, 'lhp', 'rhp', 'iuc', 'ouc'}, optional
+        Specifies whether the upper eigenvalues should be sorted.  A callable
+        may be passed that, given a eigenvalue, returns a boolean denoting
+        whether the eigenvalue should be sorted to the top-left (True). For
+        real matrix pairs, the sort function takes three real arguments
+        (alphar, alphai, beta). The eigenvalue x = (alphar + alphai*1j)/beta.
```
Comment (rgommers) [inline scipy/linalg/_decomp_qz.py]: double backticks needed around code snippets

Comment (rgommers) [inline scipy/linalg/_decomp_qz.py]: same 3 lines below

File: scipy/linalg/_decomp_qz.py [hunk 7]
```diff
@@ -160,86 +214,139 @@ def qz(A, B, output='real', lwork=None, sort=None, overwrite_a=False,
...
 
-    gges, = get_lapack_funcs(('gges',), (a1,b1))
+    output : str {'real','complex'}
+        Construct the real or complex QZ decomposition for real matrices.
+        Default is 'real'.
+    overwrite_a : bool
+        If True, the contents of A are overwritten.
+    overwrite_b : bool
+        If True, the contents of B are overwritten.
+    check_finite : boolean
```
Comment (rgommers) [inline scipy/linalg/_decomp_qz.py]: ``bool`` and for all parameters except A and B, add ``, optional``

**Comment by rgommers** [comment]
WIth current setup in numpy-vendor, there are four tests that crash: ``test_cef, test_iuc, test_lhp, test_ref``. Both with SSE2 and SSE3 builds, on Python 2.7. Didn't check other Python versions, but very likely will be the same issue. All looks like:

    .............................wine: Unhandled page fault on read access to 0xffffffff at address 0x117e5d5 (thread 0046), starting debugger...
    Application tried to create a window, but no driver could be loaded.
    Make sure that your X server is running and that $DISPLAY is set correctly.
    Unhandled exception: page fault on read access to 0xffffffff in 32-bit code (0x0117e5d5).
    Register dump:
     CS:0073 SS:007b DS:007b ES:007b FS:0033 GS:003b
     EIP:0117e5d5 ESP:0042b92c EBP:00000001 EFLAGS:00010202(  R- --  I   - - - )
     EAX:0042c0a4 EBX:0042c09c ECX:00000001 EDX:00000040
     ESI:0042c084 EDI:00000004
    Stack dump:
    0x0042b92c:  00000006 00000040 0042c09c 00000010
    0x0042b93c:  0042c09c 01c600c8 00000008 0117ee2f
    0x0042b94c:  0042c07c 01c600c0 00000002 00000003
    0x0042b95c:  00000001 00000002 0042c0bc 00000003
    0x0042b96c:  00000003 0117ed10 0042c07c 011288c1
    0x0042b97c:  00000001 00000003 00000000 3ff00000
    Backtrace:
    =>0 0x0117e5d5 in _flapack.pyd (+0x3be5d5) (0x00000001)
    0x0117e5d5: 
    Modules:
    Module  Address                 Debug info      Name (84 modules)
    PE        440000-  449000       Deferred        _multiprocessing.pyd
    PE        d10000-  db9000       Deferred        mtrand.pyd
    PE        dc0000- 1516000       Stabs           _flapack.pyd
    ...

**Comment by rgommers** [comment]
Will try again when I have https://github.com/numpy/numpy/wiki/Mingw-static-toolchain set up, maybe OpenBLAS works better.

File: scipy/linalg/flapack.pyf.src (hunk 1 above)
Comment (jseabold) [inline scipy/linalg/flapack.pyf.src]: Reverting.

File: scipy/linalg/tests/test_decomp.py (hunk 2 above)
Comment (jseabold) [inline scipy/linalg/tests/test_decomp.py]: Just saves some time so they aren't instantiated before each test method is run.

**Comment by jseabold** [comment]
@rgommers addressed comments and rebased. Updated the versionadded to 0.17.0.

**Comment by jseabold** [comment]
Cleaned up the remaining pep-8 issues.

**Comment by rgommers** [comment]
(adding 0.17.0 milestone to not forget to look at this again)

**Comment by rgommers** [comment]
After the last rebase I cannot reproduce the crash anymore. Tried with the SSE3 builds of Python 2.7 and 3.4. Maybe one of the lwork fixed did the trick here, I'm not sure. 

Also note that previously 3.2 seemed to be in the worst shape, but that's a don't care anymore by now.

**Comment by rgommers** [comment]
@jseabold one last rebase?

**Comment by jseabold** [comment]
Rebase was an odd one. Could probably use another once over, provided that CI test pass.

**Comment by rgommers** [comment]
@jseabold all green. What do you mean with odd rebase? Some commits could be squashed, but other than that it looks OK to me.

**Comment by jseabold** [comment]
There were some conflicts in which the diffs didn't really make sense to me, so I did my best. I can clean up the commit history a bit.

**Comment by jseabold** [comment]
Did a bit of cleanup.

**Comment by rgommers** [comment]
@jseabold can you edit the commit message of the last commit to remove ``[ci skip]``. The TravisCI tests won't run otherwise.

**Comment by jseabold** [comment]
Done.

**Comment by rgommers** [comment]
Allright, let's say a little prayer and hit the green button.....

**Comment by rgommers** [comment]
Merged, thanks Skipper!

**Comment by ricardomayerb** [comment]
Thank you so much, Skipper and Ralph! This is great!

**Comment by ilayn** [comment]
What is the status of this keyword disabling? I tried with LAPACK 3.7.0 and Python 3.6, Win 7 and 10 with scipy 0.19 and cannot get any segfaults from this and other PR/issues mentioned above. If we have already dropped the support for Python 3.2 or I can have a go at it about why it might segfaults. It might be worth having a `sort` option with regular QZ.

**Comment by rgommers** [comment]
Well, the Windows ATLAS binaries don't exist anymore, so I suspect we won't be able to reproduce the problem anymore. If it works with MKL, I'd say we're good for now - we can then deal with problems with other libs on Windows if and when they occur.

If you want to put in a new PR that enables ``sort``, please do.

**Pull request merged**